### ISSUE TITLE
API to add custom actions to state nodes

### ISSUE THREAD

**Issue by fabienjuif** [alakarteio/k-redux-factory#91]
From @EmrysMyrddin (https://github.com/alakarteio/k-ramel/issues/60)

---
It could be useful to allow users to define their own custom actions along with the ones already generated by k-redux-factory.

The actions could be then decorated with dispatch function, allowing the user to easily call it from the store

```es6
const store = createStore({
  data: simpleObject({ 
    actions: {
      load: () => ({ type: 'REQUEST_DATA' }),
      loaded: (data) => ({ type: 'DATA_LOADED', payload: { data } }),
    },
  }),
})

store.data.loaded()
store.data.load({ hello: 'world' })
```

Since we want to push the use of the redux-saga like API, we want to be able to dispatch very simple action like simple events. For this, our library can also help the user by generating actions for him. You just give a type name for your action and it generate the action creator for you. The action creators can take an optional payload that will be aded to the action.

```es6
const store = createStore({
  data: simpleObject({
    actions: {
      load: 'REQUEST_DATA',
      loaded: 'REQUEST_DATA',
      // Also allows to give a custom action creator implementation
      custom: () => ({ type: 'CUSTOM_ACTION' })
    },
  }),
},{
  listeners: [
    when(store.data.load.type, (action, store) => store.data.loaded({ hello: 'world' }))
    // Dispatch { type: 'DATA_LOADED', payload: { hello: 'world' } }
  ]
})
```

I'm not entirely sure of the API. In particular, I'm not convinced by the `store.data.load.type` to get the type name of the action. Perhaps by adding a `toString` implementation to the function ? Allowing to just do `when(store.data.loaded, (...) => ...)` ?

**Comment by fabienjuif** [comment]
From @EmrysMyrddin (https://github.com/alakarteio/k-ramel/issues/60)

---
An other idea for the generation of action creators is to give a list of action name and generate an action type based on the path in a mobx-state-tree fashion

```es6
const store = createStore({
  data: {
    nested: simpleObject({
      actions: [ 'load', 'loaded' ]
    }),
  },
}

store.data.nested.loaded({ hello: 'world' })
// Dispatch {type: 'data/nested/loaded', payload: { hello: 'world' } }
```

But I'm not entirely convince... It's difficult then to extend the list of actions to add a custom action creator function.

A solution can be to have both API. If `actions` is an object, we use the previously defined API, if it's an array, we use this convention.

**Comment by guillaumecrespel** [comment]
why define actions ? why do you think about that : 
```es6
const store = createStore({
  data: {
    nested: simpleObject(),
  },
}

store.data.nested.dispatch('loaded')({ hello: 'world' })
// Dispatch {type: 'data/nested/loaded', payload: { hello: 'world' } }
```

**Comment by fabienjuif** [comment]
@EmrysMyrddin is it something you still need after using the library more often ?

**Comment by EmrysMyrddin** [comment]
Yes, I did had the time toake a PR but I have began to make some utility functions on ICES project that implements my first API proposition.

For now, it works pretty well, the only drawback is that our actions are currently defined in a separate file from my model. So yes it could be great to integrate it directly in the model declaration.

**Comment by fabienjuif** [comment]
Do you want to open a PR or may I think of an implementation myself ?

**Comment by EmrysMyrddin** [comment]
I think I can make the PR. While using it, I had some ideas to improve our
first draft of the API.

Le lun. 11 juin 2018 à 08:42, Fabien JUIF <notifications@github.com> a
écrit :

**Comment by fabienjuif** [comment]
Should we wait this issue for the 6.0.0 ?

The questions behind that are:
 - do you think it will break the API ?
 - do you think we can merge the related PR during this week ?

**Comment by EmrysMyrddin** [comment]
I don't currently have the time to do the PR. But I don't think i will break existing API. I will just add things.
